File: CloudBoard.ApiService/Data/SortingApplication.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json;

namespace CloudBoard.ApiService.Data;

public class SortingApplication
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public string CreatedBy { get; set; } = string.Empty;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    public List<ProcessStep> ProcessSteps { get; set; } = new List<ProcessStep>();
}

public class ProcessStep
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public Guid Id { get; set; }
    public string ProcessStepId { get; set; } = string.Empty; // DE-W01-001
    public string StepCode { get; set; } = string.Empty; // 001
    public string ProcessStepName { get; set; } = string.Empty; // FKN Rougher
    public ProcessStepType StepType { get; set; } = ProcessStepType.Rougher;
    public string InfeedMaterialDescription { get; set; } = string.Empty;
    public string MainMaterialForEjection { get; set; } = string.Empty;
    public string ConflictingMaterials { get; set; } = string.Empty;
    public int Order { get; set; }
    
    // Foreign Keys
    public Guid SortingApplicationId { get; set; }
    public SortingApplication SortingApplication { get; set; } = null!;
    
    public Guid MarketSegmentId { get; set; }
    public MarketSegment MarketSegment { get; set; } = null!;
    
    public List<TargetMaterial> TargetMaterials { get; set; } = new List<TargetMaterial>();
}

public class MarketSegment
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public Guid Id { get; set; }
    public string SegmentCode { get; set; } = string.Empty; // W01
    public string SegmentName { get; set; } = string.Empty; // Mixed Light Packaging
    public string Country { get; set; } = string.Empty; // DE
    public BusinessUnit BusinessUnit { get; set; } = BusinessUnit.WasteRecycling;
    public string Description { get; set; } = string.Empty;
    public List<ProcessStep> ProcessSteps { get; set; } = new List<ProcessStep>();
    public List<TargetMaterial> TargetMaterials { get; set; } = new List<TargetMaterial>();
}

public class TargetMaterial
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public Guid Id { get; set; }
    public string MaterialCode { get; set; } = string.Empty; // PE-HD, PP, PET
    public string MaterialName { get; set; } = string.Empty; // Polyethylene High Density
    public MaterialCategory Category { get; set; } = MaterialCategory.Plastic;
    public string Color { get; set; } = string.Empty;
    public MaterialForm Form { get; set; } = MaterialForm.Rigid;
    public bool IsContaminant { get; set; } = false;
    
    // Navigation properties
    public List<ProcessStep> ProcessSteps { get; set; } = new List<ProcessStep>();
    public List<MarketSegment> MarketSegments { get; set; } = new List<MarketSegment>();
}

public enum ProcessStepType
{
    Rougher,
    Cleaner,
    Scavenger
}

public enum BusinessUnit
{
    WasteRecycling,
    MetalRecycling
}

public enum MaterialCategory
{
    Plastic,
    Paper,
    Metal,
    Glass,
    Wood,
    Textile,
    Other
}

public enum MaterialForm
{
    Rigid,
    Film,
    Flexible,
    Bottle,
    Tray,
    Other
}

File: CloudBoard.Angular/src/app/sorting/models/sorting-application.ts
// Enums
export enum ProcessStepType {
  Rougher = 'Rougher',
  Cleaner = 'Cleaner',
  Scavenger = 'Scavenger'
}

export enum BusinessUnit {
  WasteRecycling = 'WasteRecycling',
  MetalRecycling = 'MetalRecycling'
}

export enum MaterialCategory {
  Plastic = 'Plastic',
  Paper = 'Paper',
  Metal = 'Metal',
  Glass = 'Glass',
  Wood = 'Wood',
  Textile = 'Textile',
  Other = 'Other'
}

export enum MaterialForm {
  Rigid = 'Rigid',
  Film = 'Film',
  Flexible = 'Flexible',
  Bottle = 'Bottle',
  Tray = 'Tray',
  Other = 'Other'
}

// Domain Models
export interface SortingApplication {
  id: string;
  name: string;
  description: string;
  createdBy: string;
  createdAt: Date;
  updatedAt: Date;
  processSteps: ProcessStep[];
}

export interface ProcessStep {
  id: string;
  processStepId: string; // DE-W01-001
  stepCode: string; // 001
  processStepName: string; // FKN Rougher
  stepType: ProcessStepType;
  infeedMaterialDescription: string;
  mainMaterialForEjection: string;
  conflictingMaterials: string;
  order: number;
  marketSegment: MarketSegment;
  targetMaterials: TargetMaterial[];
}

export interface MarketSegment {
  id: string;
  segmentCode: string; // W01
  segmentName: string; // Mixed Light Packaging
  country: string; // DE
  businessUnit: BusinessUnit;
  description: string;
}

export interface TargetMaterial {
  id: string;
  materialCode: string; // PE-HD, PP, PET
  materialName: string; // Polyethylene High Density
  category: MaterialCategory;
  color: string;
  form: MaterialForm;
  isContaminant: boolean;
}

// Properties for SortingApplication nodes
export interface SortingApplicationProperties extends NodeProperties {
  applicationName: string;
  description: string;
  processStepCount: number;
  marketSegment: string;
  targetMaterials: string[];
}

// Base properties interface (from existing CloudBoard)
export interface NodeProperties {
  [key: string]: any;
  width?: string;
  height?: string;
}
